verilog/preprocessor/verilog_preprocess.cc

```diff
@@ -552,6 +552,99 @@ absl::Status VerilogPreprocess::HandleEndif(
   return absl::OkStatus();
 }
 
+// Handle `include directives.
+absl::Status VerilogPreprocess::HandleInclude(
+    TokenStreamView::const_iterator iter,
+    const StreamIteratorGenerator& generator) {
+  // karimtera(TODO): how to differentiate between <file> and "file"?? both are
+  // the same token, need to edit the lexer.
+  TokenStreamView::const_iterator token_iter =
+      generator();  // token_iter should now be pointing to a token containing
+                    // the file path.
+  auto file_token_iter = *token_iter;
+  if (file_token_iter->token_enum() != TK_StringLiteral) {
+    preprocess_data_.errors.push_back(
+        {**token_iter, "Expected a path to a SV file."});
+    return absl::InvalidArgumentError("Expected a path to a SV file.");
+  }
+  // currently the file path looks like "path", we need to remove "".
+  const auto& token_text = file_token_iter->text();
+  absl::string_view file_path(
+      token_text.begin() + 1,
+      token_text.size() - 2);  // removed 2 " chars surrounding the path.
+
+  bool is_absolute = !file_path.find_first_of("/");
+
+  std::string file_path_to_search = std::string(file_path);
+
+  if (auto status = verible::file::FileExists(file_path_to_search);
+      !status.ok()) {  // file doesn't exist in the current directory.
+    // if it's absolute then it's an error.
+    if (is_absolute) {
+      preprocess_data_.errors.push_back({**token_iter, "no such file found."});
+      return absl::InvalidArgumentError("ERROR: included file can't be found.");
+    }
+
+    bool found = 0;
+    for (const auto& base_dir : preprocess_info_.include_dirs) {
+      file_path_to_search = absl::StrCat(base_dir, "/", file_path);
+      auto search_status = verible::file::FileExists(file_path_to_search);
+      if (search_status.ok()) {
+        found = 1;
+        break;
+      }
+    }
+    if (!found) {
+      preprocess_data_.errors.push_back({**token_iter, "no such file found."});
+      return absl::InvalidArgumentError("ERROR: included file can't be found.");
+    }
+  }
+
+  // actually open the file.
+  std::string source_contents;
+  if (auto status =
+          verible::file::GetContents(file_path_to_search, &source_contents);
+      !status.ok()) {
+    preprocess_data_.errors.push_back(
+        {**token_iter, "ERROR: passed file can't be open"});
+    return absl::InvalidArgumentError("ERROR: passed file can't be open.");
+  }
+
+  verilog::VerilogPreprocess child_preprocessor(config_);
+  // TODO(karimtera): share the macros defined with child pp.
+  // TODO(karimtera): limit number of nested pps.
+  absl::string_view source_contents_view = source_contents;
+  preprocess_data_.child_included_content.emplace_back(std::make_unique<verible::TextStructure>(source_contents_view));
```

Comment:
Initially, I wanted `preprocess_data_.child_included_content` to be a vector of `TextStructure`, not `unique_ptr<TextStructure>`, but I got the following error when emplacing back:
```log
error: static assertion failed: result type must be constructible from value type of input range
```
Which I don't understand because a `TextStructure` is constructible from a string_view.

Reply:
Also, this way the `TextStructure` get destructed, and the preprocessor crashes on includes.

Reply:
The reason why you can't have a vector of plain TextStructures is, that the vector internally already allocate the space for a bunch of them, even before the first push_back().
But that means, it needs a default constructor for that to work.

Reply:
This is the only constructor available for the text structure.
 
https://github.com/chipsalliance/verible/blob/882481ad4b0d29d065446f80a219b99299fc7c30/common/text/text_structure.h#L237

Reply:
Yes, and that is intentional. So for you use-case you'd do what you do now, storing raw or unique_ptrs in your vector.